ncm-spma/src/main/perl/spma/yumng.pm

```diff
@@ -0,0 +1,810 @@
+	    return 0 if !defined($yum_clean);
+	}
+    }
+
+    # Clean up YUM state - worth to be thorough there
+    my $yum_clean = $self->execute_yum_command( ["yum clean expire-cache " . YUM_PLUGIN_OPTS], "expiring YUM caches", 1 );
+    if ($yum_warned) {
+        # Expiration of cache is not enough sometimes.
+        # https://bugzilla.redhat.com/show_bug.cgi?id=1151074
+        my $yum_clean = $self->execute_yum_command( ["yum clean all " . YUM_PLUGIN_OPTS], "resetting hard YUM state", 1 );
+        return 0 if !defined($yum_clean);
+    }
+
+    return 0 if !defined($yum_clean);
+    $yum_clean = $self->execute_yum_command( ["yum makecache " . YUM_PLUGIN_OPTS], "generating YUM cache", 1 );
+    return 0 if !defined($yum_clean);
+
+    my @files = glob "{/tmp/*.yumtx,/var/lib/yum/transaction*}";
+    foreach my $file (@files) {
+        if ( !unlink $file ) {
+            $self->info("unable to remove file $file: $!");
+        }
+    }
+    my @dirs = glob "/var/tmp/yum-root*";
+    foreach my $dir (@dirs) {
+        if ( !rmtree $dir) {
+            $self->info("unable to remove directory $dir: $!");
+        }
+    }
+
+    # Test whether comps/groups are sane.
+    {
+	$yum_warned = 0;
+	my $groupinfo = $self->execute_yum_command( [ "yum groupinfo core " . YUM_PLUGIN_OPTS ], "testing comps/groups sanity", 1 );
+	if ( $yum_warned ) {
+	    $self->error("Groups are not sane - core group missing. Will not continue.");
+	    return 0;
+	}
+    }
+
+    # Query metadata for version locked packages including Epoch and write versionlock.list
+    my $repodata_rpms = $self->execute_yum_command( [REPO_AVAIL_PKGS], "fetching full package list", 1 );
+    return 0 if !defined($repodata_rpms);
+
+    # Test whether locked packages are present in the metadata
+    my $repoquery_list = Set::Scalar->new( reverse(split ( /\n/, $repodata_rpms)) );
+    my $repoquery_list_noepoch = Set::Scalar->new();
+    my $locked_found           = Set::Scalar->new();
+    my $locked_found_noepoch   = Set::Scalar->new();
+    $self->info( "(" . $repoquery_list->size . " total packages)" );
+    while ( defined( my $p = $repoquery_list->each ) ) {
+        my $t = $p;
+        $t =~ s/^.*://;
+        if ( !$repoquery_list_noepoch->has($t) ) { $repoquery_list_noepoch->insert($t); }
+        if ( $wanted_pkgs_locked->has($t) ) {
+            if ( !$locked_found->has($p) ) {
+                $self->verbose("Found package $p");
+                $locked_found->insert($p);
+                $locked_found_noepoch->insert($t);
+            }
+        }
+    }
+    {
+        my $fh = CAF::FileWriter->new( YUM_PACKAGE_LIST, log => $self );
+        print $fh join ( "\n", @$locked_found );
+        $fh->close();
+        if ( $wanted_pkgs_locked->size != $locked_found->size ) {
+            $self->error( "Version-locked packages are missing from repositories - expected ", $wanted_pkgs_locked->size, ", available ", $locked_found->size );
```

Comment:
there should be only one error logged here (either on big message, or one error with the summary and e.g 1 or 2 warn with the rest of the data).
also the `missing` is for the case `$wanted_pkgs_locked->size > $locked_found->size`, not for the other inequality (although i don't know how that could occur)